src/shared/dissect-image.c

```diff
@@ -1418,6 +1418,69 @@ int dissect_image_and_warn(
         }
 }
 
+int blockdev_open_parent(dev_t devnum, int *ret) {
+        _cleanup_(sd_device_unrefp) sd_device *d = NULL;
+        const char *name, *devtype, *node;
+        sd_device *parent;
+        dev_t pn;
+        int fd, r;
+
+        assert(ret);
+
+        r = sd_device_new_from_devnum(&d, 'b', devnum);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to open device: %m");
+
+        if (sd_device_get_devname(d, &name) < 0) {
+                r = sd_device_get_syspath(d, &name);
+                if (r < 0) {
+                        log_debug_errno(r, "Device %u:%u does not have a name, ignoring: %m", major(devnum), minor(devnum));
+                        return 0;
+                }
+        }
+
+        r = sd_device_get_parent(d, &parent);
+        if (r < 0) {
+                log_debug_errno(r, "%s: not a partitioned device, ignoring: %m", name);
+                return 0;
+        }
+
+        /* Does it have a devtype? */
+        r = sd_device_get_devtype(parent, &devtype);
+        if (r < 0) {
+                log_debug_errno(r, "%s: parent doesn't have a device type, ignoring: %m", name);
+                return 0;
+        }
+
+        /* Is this a disk or a partition? We only care for disks... */
+        if (!streq(devtype, "disk")) {
+                log_debug("%s: parent isn't a raw disk, ignoring.", name);
+                return 0;
+        }
+
+        /* Does it have a device node? */
+        r = sd_device_get_devname(parent, &node);
+        if (r < 0) {
+                log_debug_errno(r, "%s: parent device does not have device node, ignoring: %m", name);
+                return 0;
+        }
+
+        log_debug("%s: root device %s.", name, node);
+
+        r = sd_device_get_devnum(parent, &pn);
+        if (r < 0) {
+                log_debug_errno(r, "%s: parent device is not a proper block device, ignoring: %m", name);
+                return 0;
+        }
+
+        fd = open(node, O_RDONLY|O_CLOEXEC|O_NOCTTY);
+        if (fd < 0)
+                return log_error_errno(errno, "Failed to open %s: %m", node);
```

Comment:
hmm, this is now a library function, so i figure it shouldn't log on its own anymore above LOG_DEBUG. hence maybe downgrade this to `log_debug_errno()` or drop it entirely? and then have the callers log about the error?